hs_core/tests/api/rest/test_folders.py

```diff
@@ -0,0 +1,113 @@
+        content = json.loads(response.content)
+        self.assertEqual(content, "Cannot remove folder")
+
+        # create a folder
+        response = self.client.put(url2, {})
+        self.assertEqual(response.status_code, status.HTTP_201_CREATED)
+
+        # should be able to create it again
+        response = self.client.put(url2, {})
+        self.assertEqual(response.status_code, status.HTTP_201_CREATED)
+
+        # list that folder: should work, should be empty
+        response = self.client.get(url2, {})
+        content = json.loads(response.content)
+        self.assertEqual(len(content['folders']), 0)
+        self.assertEqual(len(content['files']), 0)
+
+        # delete that folder: should be possible
+        response = self.client.delete(url2, {})
+        self.assertEqual(response.status_code, status.HTTP_200_OK)
+
+    def test_file_in_folder(self):
+        rtype = 'GenericResource'
+        title = 'My Test resource'
+        params = {'resource_type': rtype,
+                  'title': title,
+                  'file': ('cea.tif',
+                           open('hs_core/tests/data/cea.tif'),
+                           'image/tiff')}
+        url = '/hsapi/resource/'
+        response = self.client.post(url, params)
+        self.assertEqual(response.status_code, status.HTTP_201_CREATED)
+        content = json.loads(response.content)
+        res_id = content['resource_id']
+        self.resources_to_delete.append(res_id)
+
+        # create a folder 'foo'
+        url2 = str.format('/hsapi/resource/{}/folders/foo/', res_id)
+        response = self.client.put(url2, {})
+        self.assertEqual(response.status_code, status.HTTP_201_CREATED)
+
+        # create a folder 'foo/bar'
+        url3 = url2 + 'bar/'
+        response = self.client.put(url3, {})
+        self.assertEqual(response.status_code, status.HTTP_201_CREATED)
+
+        # put a file 'test.txt' into folder 'foo'
+        url4 = str.format('/hsapi/resource/{}/files/foo/', res_id)
+        params = {'file': ('test.txt',
+                           open('hs_core/tests/data/test.txt'),
+                           'text/plain')}
+        response = self.client.post(url4, params)
+        self.assertEqual(response.status_code, status.HTTP_201_CREATED)
+
+        resfile = ResourceFile.objects.get(file_folder='foo')
+        path = get_path(resfile, 'test.txt')
+        self.assertEqual(path,
+                         str.format("{}/data/contents/foo/test.txt",
+                                    res_id))
+
+        # list that folder: should work, should contain one file
+        response = self.client.get(url2, {})
+        content = json.loads(response.content)
+        self.assertEqual(len(content['folders']), 1)
+        self.assertEqual(content['folders'][0], 'bar')
+        # these do not work due to a bug in list_folder
+        # self.assertEqual(len(content['files']), 1)
+        # self.assertEqual(content['files'][0], 'test.txt')
```

Comment:
@alvacouch I think that is because the file test.txt did not get put to the right folder in lines 86-89. I think this line ```url4 = str.format('/hsapi/resource/{}/files/foo/', res_id)``` needs to be changed to ```url4 = str.format('/hsapi/resource/{}/folders/foo/', res_id)```. Otherwise, this file is not under the folder foo. Can you check the test folder logic to make sure the file is created under the right subfolder when you verify the file exists? Since listdir() is used by irods file browser and folder browsing and in many other places, it is very unlikely it would have this kind of basic bug.